File: frontend/js/envelope-engine.js
/**
 * PayEnvelope Envelope Engine (client-side)
 * ===========================================
 * JavaScript port of backend/app/envelope_engine.py. Deterministic
 * percentage / fixed / remainder allocation rules — no ML, no AI.
 */

const AllocationType = { PERCENTAGE: "PERCENTAGE", FIXED: "FIXED", REMAINDER: "REMAINDER" };

const DEFAULT_ALLOCATION_RULES = [
  { envelopeName: "Rent", type: AllocationType.PERCENTAGE, value: 0.30, color: "#6366F1", priority: 1 },
  { envelopeName: "Transportation", type: AllocationType.PERCENTAGE, value: 0.15, color: "#38BDF8", priority: 2 },
  { envelopeName: "Savings", type: AllocationType.PERCENTAGE, value: 0.10, color: "#10B981", priority: 3 },
  { envelopeName: "Utilities", type: AllocationType.PERCENTAGE, value: 0.05, color: "#F59E0B", priority: 4 },
  { envelopeName: "Emergency Fund", type: AllocationType.PERCENTAGE, value: 0.05, color: "#EF4444", priority: 5 },
  { envelopeName: "Discretionary Spending", type: AllocationType.REMAINDER, value: 0.0, color: "#8B5CF6", priority: 6 },
];

const DEFAULT_ENVELOPE_CATALOG = [
  "Rent", "Food", "Transportation", "Fuel", "Utilities", "Internet", "Electricity",
  "Water", "DSTV", "Savings", "Emergency Fund", "Health", "Insurance", "Education",
  "Parents", "Giving", "Entertainment", "Shopping", "Travel", "Miscellaneous",
];

function uuid() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

function validateRuleSet(rules) {
  const warnings = [];
  for (const r of rules) {
    if (r.type === AllocationType.PERCENTAGE && !(r.value >= 0 && r.value <= 1)) {
      throw new Error(`Percentage rule for '${r.envelopeName}' must be between 0 and 1, got ${r.value}`);
    }
    if (r.type === AllocationType.FIXED && r.value < 0) {
      throw new Error(`Fixed rule for '${r.envelopeName}' cannot be negative`);
    }
  }
  const pctTotal = rules.filter((r) => r.type === AllocationType.PERCENTAGE).reduce((s, r) => s + r.value, 0);
  const remainderRules = rules.filter((r) => r.type === AllocationType.REMAINDER);
  if (pctTotal > 1.0) warnings.push(`Percentage rules sum to ${(pctTotal * 100).toFixed(1)}%, which exceeds 100%.`);
  if (remainderRules.length > 1) warnings.push("More than one REMAINDER rule is defined; only the first will receive funds.");
  return warnings;
}

class EnvelopeEngine {
  constructor(rules) {
    this.rules = rules || DEFAULT_ALLOCATION_RULES.map((r) => ({ ...r, id: uuid() }));
    this.envelopes = {}; // name -> envelope object
    this.lastWarnings = [];
  }

  allocate(netSalary) {
    if (netSalary < 0) throw new Error("netSalary cannot be negative.");
    this.lastWarnings = validateRuleSet(this.rules);

    const ordered = [...this.rules].sort((a, b) => a.priority - b.priority);
    let remaining = netSalary;
    let remainderRule = null;

    for (const rule of ordered) {
      if (rule.type === AllocationType.REMAINDER) {
        if (!remainderRule) remainderRule = rule;
        continue;
      }
      let amount = rule.type === AllocationType.PERCENTAGE ? netSalary * rule.value : rule.value;
      amount = Math.round(Math.min(amount, Math.max(remaining, 0.0)) * 100) / 100;
      remaining = Math.round((remaining - amount) * 100) / 100;
      this._upsertEnvelope(rule.envelopeName, amount, rule.color, rule.priority);
    }

    if (remainderRule) {
      const amount = Math.round(Math.max(remaining, 0.0) * 100) / 100;
      this._upsertEnvelope(remainderRule.envelopeName, amount, remainderRule.color, remainderRule.priority);
    }

    return this.envelopes;
  }

  _upsertEnvelope(name, amount, color, priority) {
    let env = this.envelopes[name];
    if (!env) {
      env = { id: uuid(), name, balance: 0, allocated: 0, color, priority, locked: false, archived: false, recurring: false };
      this.envelopes[name] = env;
    }
    env.allocated = amount;
    env.balance = amount;
  }

  createEnvelope(name, color = "#6366F1", priority = 99) {
    if (this.envelopes[name]) throw new Error(`Envelope '${name}' already exists.`);
    const env = { id: uuid(), name, balance: 0, allocated: 0, color, priority, locked: false, archived: false, recurring: false };
    this.envelopes[name] = env;
    return env;
  }

  renameEnvelope(oldName, newName) {
    const env = this._require(oldName);
    env.name = newName;
    this.envelopes[newName] = env;
    delete this.envelopes[oldName];
  }

  deleteEnvelope(name) {
    const env = this._require(name);
    if (env.locked) throw new Error(`Envelope '${name}' is locked; unlock before deleting.`);
    delete this.envelopes[name];
  }

  archiveEnvelope(name) { this._require(name).archived = true; }
  unarchiveEnvelope(name) { this._require(name).archived = false; }
  lockEnvelope(name) { this._require(name).locked = true; }
  unlockEnvelope(name) { this._require(name).locked = false; }

  mergeEnvelopes(sourceName, targetName) {
    const source = this._require(sourceName);
    const target = this._require(targetName);
    target.balance = Math.round((target.balance + source.balance) * 100) / 100;
    target.allocated = Math.round((target.allocated + source.allocated) * 100) / 100;
    delete this.envelopes[sourceName];
    return target;
  }

  splitEnvelope(name, newName, fraction) {
    if (!(fraction > 0 && fraction < 1)) throw new Error("fraction must be strictly between 0 and 1.");
    const source = this._require(name);
    const movedBalance = Math.round(source.balance * fraction * 100) / 100;
    const movedAllocated = Math.round(source.allocated * fraction * 100) / 100;
    source.balance = Math.round((source.balance - movedBalance) * 100) / 100;
    source.allocated = Math.round((source.allocated - movedAllocated) * 100) / 100;
    const newEnv = { id: uuid(), name: newName, balance: movedBalance, allocated: movedAllocated, color: source.color, priority: source.priority, locked: false, archived: false, recurring: false };
    this.envelopes[newName] = newEnv;
    return [source, newEnv];
  }

  transfer(fromName, toName, amount) {
    const source = this._require(fromName);
    const target = this._require(toName);
    if (source.locked) throw new Error(`Envelope '${source.name}' is locked and cannot be spent from.`);
    if (amount < 0) throw new Error("Transfer amount must be non-negative.");
    source.balance = Math.round((source.balance - amount) * 100) / 100;
    target.balance = Math.round((target.balance + amount) * 100) / 100;
  }

  spend(name, amount) {
    const env = this._require(name);
    if (env.locked) throw new Error(`Envelope '${env.name}' is locked and cannot be spent from.`);
    env.balance = Math.round((env.balance - amount) * 100) / 100;
  }

  deposit(name, amount) {
    const env = this._require(name);
    env.balance = Math.round((env.balance + amount) * 100) / 100;
  }

  _require(name) {
    const env = this.envelopes[name];
    if (!env) throw new Error(`No envelope named '${name}'.`);
    return env;
  }
}

window.PayEnvelopeEngine = { EnvelopeEngine, AllocationType, DEFAULT_ALLOCATION_RULES, DEFAULT_ENVELOPE_CATALOG, validateRuleSet, uuid };

File: backend/app/envelope_engine.py
"""
PayEnvelope Envelope Engine
============================
Deterministic, rule-based budget allocation. No ML, no AI — every allocation
decision traces back to a percentage, a fixed amount, or a "remainder" rule
that a user configured (or a documented Nigeria-first default).

Core concepts
-------------
Envelope        A named bucket of money with a balance, a color, a priority,
                and optional lock/archive/recurring flags.
AllocationRule  Describes how a slice of net salary is assigned to an
                envelope: a fixed percentage, a fixed amount, or "remainder"
                (whatever's left after all percentage/fixed rules run).
EnvelopeEngine  Applies an ordered list of AllocationRules against a net
                salary figure to produce envelope balances, and provides the
                envelope lifecycle operations (create/rename/archive/merge/
                split/lock) requested in the product brief.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional


class AllocationType(str, Enum):
    PERCENTAGE = "PERCENTAGE"   # e.g. 0.30 of net salary
    FIXED = "FIXED"             # e.g. a flat amount regardless of salary size
    REMAINDER = "REMAINDER"     # whatever is left after all other rules run


# Nigeria-first default rule set, mirroring the brief's example allocation.
# Percentages sum to 65%; the remaining 35% is split between an explicit
# emergency-fund percentage and a REMAINDER envelope for discretionary spend.
DEFAULT_ALLOCATION_RULES: list[dict] = [
    {"envelope_name": "Rent",            "type": AllocationType.PERCENTAGE, "value": 0.30, "color": "#6366F1", "priority": 1},
    {"envelope_name": "Transportation",  "type": AllocationType.PERCENTAGE, "value": 0.15, "color": "#38BDF8", "priority": 2},
    {"envelope_name": "Savings",         "type": AllocationType.PERCENTAGE, "value": 0.10, "color": "#10B981", "priority": 3},
    {"envelope_name": "Utilities",       "type": AllocationType.PERCENTAGE, "value": 0.05, "color": "#F59E0B", "priority": 4},
    {"envelope_name": "Emergency Fund",  "type": AllocationType.PERCENTAGE, "value": 0.05, "color": "#EF4444", "priority": 5},
    {"envelope_name": "Discretionary Spending", "type": AllocationType.REMAINDER, "value": 0.0, "color": "#8B5CF6", "priority": 6},
]

DEFAULT_ENVELOPE_CATALOG = [
    "Rent", "Food", "Transportation", "Fuel", "Utilities", "Internet",
    "Electricity", "Water", "DSTV", "Savings", "Emergency Fund", "Health",
    "Insurance", "Education", "Parents", "Giving", "Entertainment",
    "Shopping", "Travel", "Miscellaneous",
]


@dataclass
class AllocationRule:
    envelope_name: str
    type: AllocationType
    value: float  # fraction (0-1) for PERCENTAGE, absolute amount for FIXED, ignored for REMAINDER
    color: str = "#6366F1"
    priority: int = 99
    id: str = field(default_factory=lambda: str(uuid.uuid4()))

    def validate(self) -> None:
        if self.type == AllocationType.PERCENTAGE and not (0 <= self.value <= 1):
            raise ValueError(f"Percentage rule for '{self.envelope_name}' must be between 0 and 1, got {self.value}")
        if self.type == AllocationType.FIXED and self.value < 0:
            raise ValueError(f"Fixed rule for '{self.envelope_name}' cannot be negative")


@dataclass
class Envelope:
    name: str
    balance: float = 0.0
    allocated: float = 0.0
    color: str = "#6366F1"
    priority: int = 99
    locked: bool = False
    archived: bool = False
    recurring: bool = False
    id: str = field(default_factory=lambda: str(uuid.uuid4()))

    def spend(self, amount: float) -> None:
        if self.locked:
            raise PermissionError(f"Envelope '{self.name}' is locked and cannot be spent from.")
        if amount < 0:
            raise ValueError("Spend amount must be non-negative.")
        self.balance = round(self.balance - amount, 2)

    def deposit(self, amount: float) -> None:
        if amount < 0:
            raise ValueError("Deposit amount must be non-negative.")
        self.balance = round(self.balance + amount, 2)


def validate_rule_set(rules: list[AllocationRule]) -> list[str]:
    """Return a list of human-readable warnings (empty list = clean)."""
    warnings: list[str] = []
    for r in rules:
        r.validate()

    pct_total = sum(r.value for r in rules if r.type == AllocationType.PERCENTAGE)
    remainder_rules = [r for r in rules if r.type == AllocationType.REMAINDER]

    if pct_total > 1.0:
        warnings.append(f"Percentage rules sum to {pct_total * 100:.1f}%, which exceeds 100%.")
    if len(remainder_rules) > 1:
        warnings.append("More than one REMAINDER rule is defined; only the first will receive funds.")
    return warnings


class EnvelopeEngine:
    """Applies AllocationRules to a net salary figure and manages envelope
    lifecycle operations (create, rename, delete, archive, merge, split, lock)."""

    def __init__(self, rules: Optional[list[AllocationRule]] = None):
        self.rules: list[AllocationRule] = rules or [AllocationRule(**r) for r in DEFAULT_ALLOCATION_RULES]
        self.envelopes: dict[str, Envelope] = {}

    # -- Allocation ---------------------------------------------------

    def allocate(self, net_salary: float) -> dict[str, Envelope]:
        """Distribute `net_salary` across envelopes per the active rule set.
        Fixed and percentage rules are applied first (sorted by priority),
        then a single REMAINDER rule (if any) absorbs whatever is left."""
        if net_salary < 0:
            raise ValueError("net_salary cannot be negative.")

        warnings = validate_rule_set(self.rules)  # noqa: F841 (surfaced via .last_warnings if needed)
        self.last_warnings = warnings

        ordered = sorted(self.rules, key=lambda r: r.priority)
        remaining = net_salary
        remainder_rule: Optional[AllocationRule] = None

        for rule in ordered:
            if rule.type == AllocationType.REMAINDER:
                if remainder_rule is None:
                    remainder_rule = rule
                continue

            amount = (net_salary * rule.value) if rule.type == AllocationType.PERCENTAGE else rule.value
            amount = round(min(amount, max(remaining, 0.0)), 2)
            remaining = round(remaining - amount, 2)
            self._upsert_envelope(rule.envelope_name, amount, rule.color, rule.priority)

        if remainder_rule is not None:
            amount = round(max(remaining, 0.0), 2)
            self._upsert_envelope(remainder_rule.envelope_name, amount, remainder_rule.color, remainder_rule.priority)

        return self.envelopes

    def _upsert_envelope(self, name: str, amount: float, color: str, priority: int) -> None:
        env = self.envelopes.get(name)
        if env is None:
            env = Envelope(name=name, color=color, priority=priority)
            self.envelopes[name] = env
        env.allocated = amount
        env.balance = amount

    # -- Lifecycle operations ------------------------------------------

    def create_envelope(self, name: str, color: str = "#6366F1", priority: int = 99) -> Envelope:
        if name in self.envelopes:
            raise ValueError(f"Envelope '{name}' already exists.")
        env = Envelope(name=name, color=color, priority=priority)
        self.envelopes[name] = env
        return env

    def rename_envelope(self, old_name: str, new_name: str) -> None:
        env = self._require(old_name)
        env.name = new_name
        self.envelopes[new_name] = self.envelopes.pop(old_name)

    def delete_envelope(self, name: str) -> None:
        env = self._require(name)
        if env.locked:
            raise PermissionError(f"Envelope '{name}' is locked; unlock before deleting.")
        del self.envelopes[name]

    def archive_envelope(self, name: str) -> None:
        self._require(name).archived = True

    def unarchive_envelope(self, name: str) -> None:
        self._require(name).archived = False

    def lock_envelope(self, name: str) -> None:
        self._require(name).locked = True

    def unlock_envelope(self, name: str) -> None:
        self._require(name).locked = False

    def merge_envelopes(self, source_name: str, target_name: str) -> Envelope:
        source = self._require(source_name)
        target = self._require(target_name)
        target.balance = round(target.balance + source.balance, 2)
        target.allocated = round(target.allocated + source.allocated, 2)
        del self.envelopes[source_name]
        return target

    def split_envelope(self, name: str, new_name: str, fraction: float) -> tuple[Envelope, Envelope]:
        if not (0 < fraction < 1):
            raise ValueError("fraction must be strictly between 0 and 1.")
        source = self._require(name)
        moved_balance = round(source.balance * fraction, 2)
        moved_allocated = round(source.allocated * fraction, 2)
        source.balance = round(source.balance - moved_balance, 2)
        source.allocated = round(source.allocated - moved_allocated, 2)
        new_env = Envelope(name=new_name, balance=moved_balance, allocated=moved_allocated,
                            color=source.color, priority=source.priority)
        self.envelopes[new_name] = new_env
        return source, new_env

    def transfer(self, from_name: str, to_name: str, amount: float) -> None:
        source = self._require(from_name)
        target = self._require(to_name)
        source.spend(amount)
        target.deposit(amount)

    def _require(self, name: str) -> Envelope:
        if name not in self.envelopes:
            raise KeyError(f"No envelope named '{name}'.")
        return self.envelopes[name]
